src/resolve/resolved-dns-browse-services.c

```diff
@@ -0,0 +1,776 @@
+         * The intervals between successive queries MUST increase by at least a
+         * factor of two. When the interval between queries reaches or exceeds
+         * 60 minutes,perform subsequent queries at a steady-state rate of one
+         * query per hour */
+
+        sb->delay = sb->delay < 2048 ? sb->delay * 2 : 3600;
+
+        SET_FLAG(sb->flags, SD_RESOLVED_NO_CACHE, true);
+
+        r = event_reset_time_relative(
+                        sb->manager->event,
+                        &sb->schedule_event,
+                        CLOCK_BOOTTIME,
+                        (sb->delay * USEC_PER_SEC),
+                        /* accuracy= */ 0,
+                        mdns_next_query_schedule,
+                        sb,
+                        /* priority= */ 0,
+                        "mdns-next-query-schedule",
+                        /* force_reset= */ true);
+        if (r < 0)
+                return log_error_errno(r, "Failed to reset event time for next query schedule: %m");
+
+        TAKE_PTR(q);
+
+        return 0;
+}
+
+void dns_browse_services_restart(Manager *m) {
+        int r;
+
+        if (!(m && m->dns_service_browsers))
+                return;
+
+        DnsServiceBrowser *sb;
+
+        HASHMAP_FOREACH(sb, m->dns_service_browsers) {
+                sb->delay = 0;
+
+                r = event_reset_time_relative(
+                                sb->manager->event,
+                                &sb->schedule_event,
+                                CLOCK_BOOTTIME,
+                                (sb->delay * USEC_PER_SEC),
+                                /* accuracy= */ 0,
+                                mdns_next_query_schedule,
+                                sb,
+                                /* priority= */ 0,
+                                "mdns-next-query-schedule",
+                                /* force_reset= */ true);
+
+                if (r < 0) {
+                        log_error_errno(r,
+                                        "Failed to reset mDNS service subscriber event "
+                                        "for service browser: %m");
+                }
+        }
+}
+
+int dns_subscribe_browse_service(
+                Manager *m, sd_varlink *link, const char *domain, const char *type, int ifindex, uint64_t flags) {
+
+        _cleanup_(dns_service_browser_unrefp) DnsServiceBrowser *sb = NULL;
+        _cleanup_(dns_question_unrefp) DnsQuestion *question_idna = NULL, *question_utf8 = NULL;
+        int r;
+
+        assert(m);
+        assert(link);
+
+        if (ifindex < 0)
+                return sd_varlink_error_invalid_parameter_name(link, "ifindex");
+
+        if (isempty(type))
+                type = NULL;
+        else if (!dnssd_srv_type_is_valid(type))
+                return sd_varlink_error_invalid_parameter_name(link, "type");
+
+        if (isempty(domain))
+                domain = "local";
+
+        r = dns_name_is_valid(domain);
```

Comment:
maybe do this inside an `else` branch, after all if we set the domain to `local` we know its valid...